File: java/RepeatedDNASequences.java
/*
 * Filename: RepeatedDNASequences.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 15 Jul 2015 10:23:33 PM EDT
 */

public class Solution {
	public List<String> findRepeatedDnaSequences(String s) {
		List<String> res = new ArrayList<String>();

		if (s.length() <= 10) {
			return res;
		}

		Map<Character, Integer> map = new HashMap<Character, Integer>();
		map.put('A', 0);
		map.put('C', 1);
		map.put('G', 2);
		map.put('T', 3);

		HashMap<Integer, Integer> dupMap = new HashMap<Integer, Integer>();

		int hash = 0;
		for (int i = 0; i < s.length(); i++) {
			if (i < 9) {
				hash = (hash << 2) + map.get(s.charAt(i));
			} else {
				hash = (hash << 2) + map.get(s.charAt(i));
				hash = hash & (1 << 20) - 1;

				if (dupMap.containsKey(hash)) {
					if (dupMap.get(hash) == 1) {
						res.add(s.substring(i-9, i+1));
						dupMap.put(hash, 2);
					}
				} else {
					dupMap.put(hash, 1);
				}
			}
		}

		return res;
	}
}

File: cpp/repeated_dna_sequences.cpp
/*
 * Filename: repeated_dna_sequences.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Tue 17 Nov 2015 12:12:36 AM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution {
	public:
		vector<string> findRepeatedDnaSequences(string s) {
			vector<string> res;

			if (s.length() <= 10)
			{
				return res;
			}

			map<int, int> dupmap;
			map<char, int> map;
			map['A'] = 0;
			map['C'] = 1;
			map['G'] = 2;
			map['T'] = 3;

			int hash = 0;
			for (int i = 0; i < s.length(); i++)
			{
				if (i < 9)
				{
					hash = (hash << 2) + map[s[i]];
				}
				else
				{
					hash = (hash << 2) + map[s[i]];
					hash = hash & (1 << 20) - 1;

					if (dupmap.find(hash) != dupmap.end())
					{
						if (dupmap[hash] == 1)
						{
							res.push_back(s.substr(i-9, 10));
							dupmap[hash] = 2;
						}
					}
					else
					{
						dupmap[hash] = 1;
					}
				}
			}

			return res;
		}
};
